oeps/oep-0016.rst

```diff
@@ -0,0 +1,195 @@
+features can be built following this best practice. In edX platform, the core
+templates will be parameterized to support the option of using Bootstrap. With
+this option chosen, the page skeleton (header, footer, navigation) will
+automatically render in a mobile-first design using Bootstrap components.
+
+New Bootstrap themes will be implemented in the `edX UI Toolkit`_. These themes
+will be provided as Sass partial files that can be included into any project's
+Sass file. The files will be as follows:
+* A partial defining the standard edX fonts, layouts, margins etc
+* An edx.org brand partial which defines the edX colors and background images
+* An Open edX partial that defines the community colors and images
+
+An Open edX site can provide its own brand colors and images while still
+importing the standard fonts, layouts, margins etc. If desired, even those
+rules can be overridden to provide full flexibility in theming.
+
+One important consideration is that as of April 2017, Bootstrap 4 is still
+in alpha. Based upon the proof-of-concept, this does not seem to be an issue
+as the functionality is very complete, and the beta release is imminent. As
+this is an open source project, it is straightforward to track the work, and
+to provide fixes as necessary.
+
+Backward Compatibility
+======================
+
+The approach taken in the reference implementation was to introduce a new
+v3 style, thus leaving all v1 and v2 styles unaffected. This allows pages
+to be converted one at a time.
+
+An investigation was performed to see whether Bootstrap components could be used
+on non-Bootstrap pages, but the conclusion was that the conflicts made this more
+trouble than it was worth. The recommendation is to convert pages in one shot
+which will be quicker and will have fewer issues.
+
+Reference Implementation
+========================
+
+In order to prove the viability of Bootstrap 4, a proof-of-concept was built
+that converted three different pages from the LMS:
+
+* The course bookmarks page which was a simple first page to target
+* The main courseware page which contains a number of custom styling rules
+* The new course outline page which is a typical `edX Pattern Library`_ page
+
+The following work was done:
+
+* A new lms-main-v3.scss file was introduced that used Bootstrap
+* An edX Bootstrap theme was implemented and included as a partial
+* The three pages were switched over to use the v3 styling
+* A "uses_bootstrap" context parameter enables a template to add Bootstrap JS
+* The header was reimagined with Bootstrap components
+* The page styles were updated to use Bootstrap mixins and theme variables
+
+For the courseware, a new lms-course-v3.scss was added
+
+* This file included all of the course-specific Sass partials
+* Any global-level rules were excluded to allow Bootstrap's styles to be used
+
+See the `Bootstrap Proof-of-Concept PR`_ for more details, including screenshots
```

Comment:
I'd love to talk to you more about theming, as I'm only familiar with comprehensive theming in edX platform. For that, there is already a separate Sass file per site, so I was thinking that each file would just include its own Bootstrap theme partial. Let's get together with @douglashall, @gsong, and other interested theming folks to come up with a consistent approach.